tactics/hints.ml

```diff
@@ -595,7 +598,9 @@ struct
 
   let matches_modes sigma args modes =
     if List.is_empty modes then true
-    else List.exists (matches_mode sigma args) modes
+    else
+      List.exists (matches_mode sigma args) modes ||
+      raise ModeMatchFailure
```

Comment:
This function always returns true or raises, I guess it should return unit instead

Reply:
Yes, it will be cleaner rewriting this with an error monad instead

Reply:
The logic has been changed to use an algebraic with_mode type as you suggested.